### ISSUE TITLE
[Controller Request]: Flydigi APEX and Flydigi Apex 2

### ISSUE THREAD

**Issue by blackver** [ndeadly/MissionControl#457]
### Controller Name

Flydigi APEX

### Current Controller Behaviour

Controller is not successful in pairing with the console

### Supported Features

- [X] Rumble/Vibration
- [X] Motion Controls
- [ ] Player Number Indicator
- [X] Battery Level Reporting

### Additional Details

the official website link to the product : https://www.flydigi.com/en/apex/

**Comment by ndeadly** [comment]
The website for this controller mentions a 2.5G receiver. Is actually possible to connect it to other devices without that? Assuming you have done everything else correctly, a controller that doesn't pair with the console at all usually means the hardware isn't compatible/supported.

**Comment by blackver** [comment]
the controller can use Bluetooth to connect with other devices.

**Comment by ndeadly** [comment]
And you activate this pairing mode at the `Change Grip/Order` screen and nothing happens? Or do you see a pairing notification of some kind?

**Comment by blackver** [comment]
yes, nothing happens on Change Grip/Order screen. maybe I can try it again when i am at home today, last time I tried on V2. I will try on my OLED this time. So basically the Mission Control can capture almost all bluetooth controllers?

**Comment by ndeadly** [comment]
> So basically the Mission Control can capture almost all bluetooth controllers?

Yes, even if support for a controller hasn't been explicitly added to mission control for mapping buttons, any compatible bluetooth controller should still pair successfully. If it doesn't, likely explanations are:
- Incorrect pairing procedure. Eg. wrong pairing mode, user on the wrong menu etc.
- The controller advertises a device class other than on of several supported HID device types and is rejected by checks introduced on more recent firmwares.
- The controller uses bluetooth legacy pincode pairing, and uses a pincode other than the defauly `0000`.
- Incompatible hardware. This can be a cheap out-of-spec bluetooth chipset, a bluetooth LE controller, or one that doesn't support bluetooth at all.

We can potentially get some more information via logging some of the pairing events, but best to rule out the obvious things first. In my experience the logs contain little information of value in most cases.

**Comment by blackver** [comment]
for points number 1 and 3 I can re-check it today. for point 2, how can I check it?

**Comment by ndeadly** [comment]
> for points number 1 and 3 I can re-check it today. for point 2, how can I check it?

Either through logging (needs a debug build of mission control) or looking at device class properties when connected to a PC. Alternately, you could try pairing on a pre-12.0.0 firmware before the device class checks were added.

**Comment by blackver** [comment]
sorry I don't have pre-12.0.0 firmware switch cfw. so i will check it when connected to PC,.

**Comment by Elson-creator** [comment]
This joystick uses bluetooth LE technology, but if you connect the 2.4G wireless adapter, you can use it natively on the Switch.

**Comment by ndeadly** [comment]
That makes sense, thanks for the clarification. You mean to say it emulates a Switch controller via the USB adapter? Or do you need sys-con running?

**Comment by Elson-creator** [comment]
Just connect the USB adapter, if it doesn't work, activate sys-con.

**Comment by ndeadly** [comment]
Well, it's either one or the other. I imagine sys-con is necessary since the product page makes no mention of Switch support. This is not what I would refer to as being able to use "natively"

**Comment by Elson-creator** [comment]
Although the manufacturer's page says it doesn't work mine works with the UBS adapter without needing any homebrew controller and recognized as Xbox controller.

**Comment by blackver** [comment]
tested using the usb wireless receiver : not detected with or without mission control
tested android mode : not detected with or without mission control
tested bluetooth mode : not detected with or without mission control
Controller : Flydigi APEX

**Comment by ndeadly** [comment]
@blackver this is to be expected if what @Elson-creator is correct about the controllers using Bluetooth LE. You will need to wait until I have finished implementing LE support before this controller can be used via Bluetooth. Mission Control also doesn't support USB currently so you shouldn't expect it to work with the adapter either.  You might have some luck with sys-con for USB connections via the wireless adapter.

**Comment by blackver** [comment]
i see.. i will wait for the Bluetooth LE implementation then. Thank you.

**Comment by ndeadly** [comment]
No problem. I will follow up in this thread when it's ready in order to correctly recognise and map this controller.

**Comment by teguhsatria892** [comment]
apex 2 support with sys con using USB adapter, but I need Bluetooth connection, I wont plug and unplug my adapter from my PC to switch
